### PR TITLE
Atoms can now be stationloving

### PR DESCRIPTION
**Pull request by coiax**
Stationloving objects are objects that love the station so much, they
will relocate back to the station if spaced or if destroyed.

Yes, this does sound a lot like the behaviour of the nuke disk. Except
made more generic.

Currently the only stationloving object is the nuke disk.

### CODE REVIEW

File: code/controllers/subsystem/processing/inbounds.dm [hunk 1]
```diff
@@ -0,0 +1,35 @@
+var/datum/subsystem/objects/SSinbounds
```
Comment (Cyberboss) [inline code/controllers/subsystem/processing/inbounds.dm]: At least make it a processing subsystem and increase it's wait

**Comment by Cyberboss** [comment]
I'm just going to say this feels like subsystem creep

File: code/controllers/subsystem/processing/inbounds.dm (hunk 1 above)
Comment (coiax) [inline code/controllers/subsystem/processing/inbounds.dm]: if I made it processing subsystem, I would have to put stuff in process() on all objects, and then put notes for people to call it if they wanted it to act like it, and it was messy.

File: code/controllers/subsystem/processing/inbounds.dm [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	//cache for sanic speed (lists are references anyways)
+	var/list/currentrun = src.currentrun
+
+	while(currentrun.len)
+		var/atom/thing = currentrun[currentrun.len]
+		currentrun.len--
+		if(thing)
+			thing.check_in_bounds(wait)
+		else
+			SSobj.processing -= thing
```
Comment (Cyberboss) [inline code/controllers/subsystem/processing/inbounds.dm]: I SMELL COPY PASTA

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/processing/inbounds.dm (hunk 2 above)
Comment (coiax) [inline code/controllers/subsystem/processing/inbounds.dm]: Yes. Because I need the processing subsystem,  but for it to call a different proc.

**Comment by Shadowlight213** [comment]
Cyberboss has a point about subsystem creep.

**Comment by Incoming5643** [comment]
I don't see anything but the disk ever using this and even the disk should get a better solution than what it currently has.

**Comment by optimumtact** [comment]
Unless @MrStonedOne indicates there is some pressing need to keep the number of subsystems low, there is no such thing as subystem creep

**Comment by Cyberboss** [comment]
Its not that I care about keeping it low. It just seems like we're shooting a fly with a machine gun here

File: code/controllers/subsystem/processing/inbounds.dm (hunk 2 above)
Comment (Cyberboss) [inline code/controllers/subsystem/processing/inbounds.dm]: >SSobj

Comment (coiax) [inline code/controllers/subsystem/processing/inbounds.dm]: > it to call a different proc

**Comment by coiax** [comment]
Any object can be made stationloving with this system, which may or may not be used in the future. That's the point. I do the refactor now, to enable the easy changes later.

**Comment by Kromgar** [comment]
Station loving Sky Bulge toss it out an airlock and end up in a random part of the station.

**Comment by bobdobbington** [comment]
Yeah, this could also be useful for admin events

**Comment by coiax** [comment]
I was considering making His Grace stationfriendly, but that's for another PR.

File: code/controllers/subsystem/processing/inbounds.dm (hunk 2 above)
Comment (MrStonedOne) [inline code/controllers/subsystem/processing/inbounds.dm]: you don't understand, you need to finish changing this over and make it remove from this subsystem and not ssobj

**Comment by MrPerson** [comment]
Why does this need to be a processing subystem? You can just hook onto Moved() and if newloc.z != oldloc.z, test for station loving objects then.

**Comment by Arianya** [comment]
With this implemented it'd probably be worth doing with objects that can be traitor objectives. I.e. HoS' laser gun, RD's teleport armour, etc.

Sucks to get fucked out of your greentext because someone fucked off into space/got qdel'd by the singulo.

**Comment by RemieRichards** [comment]
expand it to atoms
make the clown stationloving
space his body? <i>he's still here</i>

**Comment by XDTM** [comment]
Throw him into the supermatter until he regrets being stationloving

**Comment by Cyberboss** [comment]
I highly agree with @MrPerson

**Comment by coiax** [comment]
Hooking onto Moved would be heavy. Even if it's just a check to see if a var is null or not, that's still a check every time something moves.

**Comment by coiax** [comment]
So I'm gonna expand it to all atom movables, but be warned that the nuke disk works fine when relocating because it has no HP to lose. But other things that qdel() when damaged will end up in /WEIRD/ states. Caveat emptor.

**Comment by Cyberboss** [comment]
Add a check for the flag in movable/Destroy. It's not THAT expensive.

Also, what you should do is add a ZChanged proc in movable/Moved that's called whenever the atoms z changes.

**Comment by coiax** [comment]
Except I've left the specifics of what classes as "out of bounds" overridable. Right now we're just going with z levels, but it could also check to whether it's in a pressurised environment, or is out of view of people.

File: code/game/atoms_movable.dm [hunk 3]
```diff
@@ -420,3 +439,78 @@
 	else if (!on && floating)
 		animate(src, pixel_y = initial(pixel_y), time = 10)
 		floating = FALSE
+
+/* Stationloving
+*
+* An stationloving atom will always teleport back to the station
```
Comment (RemieRichards) [inline code/game/atoms_movable.dm]: "An stationloving"

**Comment by coiax** [comment]
Some testing:
- Stationloving mobs relocate fine, but the Destroy() immortality doesn't work
- A made a stationloving closet, Destroy() empties it, but it does respawn

File: code/game/atoms_movable.dm [hunk 4]
```diff
@@ -420,3 +439,78 @@
 	else if (!on && floating)
 		animate(src, pixel_y = initial(pixel_y), time = 10)
 		floating = FALSE
+
+/* Stationloving
+*
+* An stationloving atom will always teleport back to the station
+* if it ever leaves the station z-levels or Centcom. It will also,
+* when Destroy() is called, will teleport to a random turf on the
```
Comment (RemieRichards) [inline code/game/atoms_movable.dm]: remove "will" here.

**Comment by Cyberboss** [comment]
Mobs return QDEL_HINT_HARDDEL and ignore hints below that

**Comment by Jalleo** [comment]
Could we make it possible for things to be mining loving or derelict loving? Or having multiple Z's that it can be at?

**Comment by coiax** [comment]
Yes. Although not with simple var editing, you'd need to override the `in_bounds` proc, or modify it so it could be configured with flags.

**Comment by Iamgoofball** [comment]
# stations and objects shouldnt be together it's not NATURAL

**Comment by optimumtact** [comment]
@MrPerson why would you add a check to every single move in the game for every single object when the change conditions happen infrequently and live on very few objects.

This is one of those cases where polling every few minutes is less overhead.

**Comment by MrPerson** [comment]
I don't think 1-3 object variable checks per move is going to break the bank, and there are other objects that will want to do special behavior on z changes anyway like meteors, the singulo, and certain mobs. Also the poll requires a get_turf() call for each object tested so any argument based on speed is pretty much out the door on proccall overhead alone. Not even getting to the fact that it's another reference that needs to be cleaned up in Destroy().

Having said that, the design of "run this verification test every so often" is more expansive and overall fine since it means more than just the nuke disk checking its z. So whatever, polling is fine I guess.

File: code/controllers/subsystem/processing/inbounds.dm [hunk 5]
```diff
@@ -0,0 +1,32 @@
+var/datum/controller/subsystem/objects/SSinbounds
+/datum/controller/subsystem/inbounds
```
Comment (Cyberboss) [inline code/controllers/subsystem/processing/inbounds.dm]: COPYFUCKINGPASTA!!!!!

Comment (coiax) [inline code/controllers/subsystem/processing/inbounds.dm]: wew.

**Pull request merged**